File: app/src/main/java/com/tejasbhitle/gallery/model/MediaModel.java
package com.tejasbhitle.gallery.model;

import android.app.Activity;
import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.tejasbhitle.gallery.R;

import java.io.File;
import java.net.URLConnection;
import java.util.List;

import androidx.annotation.NonNull;

public class MediaModel extends AbstractItem<MediaModel, MediaModel.ViewHolder>
        implements Parcelable {

    private final static int TYPE_ID = 683881;
    private static final String TAG = "MediaModel";
    private File file;

    public MediaModel(File file) {
        this.file = file;
    }

    public File getFile() {
        return file;
    }

    public void setFile(File file) {
        this.file = file;
    }

    public boolean isVideoFile() {
        try {

            String mimeType = URLConnection.guessContentTypeFromName(this.file.getPath());
            return mimeType != null && mimeType.startsWith("video");
        }catch (StringIndexOutOfBoundsException e){
            Log.e(TAG,file.getPath());
            e.printStackTrace();
        }
        return false;
    }

    @NonNull
    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    @Override
    public int getType() {
        return TYPE_ID;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.list_item_media;
    }

    public class ViewHolder extends FastAdapter.ViewHolder<MediaModel>{

        ImageView media_image;
        Context context;

        public ViewHolder(View view){
            super(view);
            context = view.getContext();
            media_image = view.findViewById(R.id.media_image);


            DisplayMetrics displayMetrics = new DisplayMetrics();
            ((Activity)context).getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);

            int width = displayMetrics.widthPixels;
            int factor = 4;
            media_image.getLayoutParams().height = width/factor;

        }

        @Override
        public void bindView(MediaModel item, List<Object> payloads) {
            if(item.isVideoFile()){

                Glide.with(context)
                        .load(item.file)
                        .apply(new RequestOptions().fitCenter().centerCrop())
                        .thumbnail(0.1f)
                        .into(media_image);
            }
            else {
                Glide.with(context)
                        .load(item.file)
                        .apply(new RequestOptions().fitCenter().centerCrop())
                        .into(media_image);
            }
        }

        @Override
        public void unbindView(MediaModel item) {

        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeSerializable(this.file);
    }

    protected MediaModel(Parcel in) {
        this.file = (File) in.readSerializable();
    }

    public static final Parcelable.Creator<MediaModel> CREATOR = new Parcelable.Creator<MediaModel>() {
        @Override
        public MediaModel createFromParcel(Parcel source) {
            return new MediaModel(source);
        }

        @Override
        public MediaModel[] newArray(int size) {
            return new MediaModel[size];
        }
    };
}

File: app/src/main/java/com/tejasbhitle/gallery/model/MediaModel.kt
package com.tejasbhitle.gallery.model

import android.app.Activity
import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.DisplayMetrics
import android.util.Log
import android.view.View
import android.widget.ImageView

import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.items.AbstractItem
import com.tejasbhitle.gallery.R

import java.io.File
import java.net.URLConnection

class MediaModel : AbstractItem<MediaModel, MediaModel.ViewHolder>, Parcelable {

    constructor(file: File?) {
        this.file = file
    }

    var file: File? = null
    private val TYPE_ID = 683881
    private val TAG = "MediaModel"

    val isVideoFile: Boolean
        get() {
            try {
                val mimeType = URLConnection.guessContentTypeFromName(this.file!!.path)
                return mimeType != null && mimeType.startsWith("video")
            } catch (e: StringIndexOutOfBoundsException) {
                Log.e(TAG, file!!.path)
                e.printStackTrace()
            }
            return false
        }

    val mimeType: String
        get() = if (isVideoFile) "video/*" else "image/*"

    val fileSize: String
        get() {
            var fileLen = file!!.length()
            if (fileLen < 1024)
                return fileLen.toString() + " b"
            fileLen /= 1024
            if (fileLen < 1024)
                return fileLen.toString() + " Kb"
            fileLen /= 1024
            if (fileLen < 1024)
                return fileLen.toString() + " Mb"
            fileLen /= 1024
            return fileLen.toString() + " Gb"
        }

    override fun getViewHolder(v: View): ViewHolder  = ViewHolder(v)

    override fun getType(): Int = TYPE_ID

    override fun getLayoutRes(): Int = R.layout.list_item_media


    inner class ViewHolder(view: View) : FastAdapter.ViewHolder<MediaModel>(view) {

        private val media_image: ImageView
        private val context: Context

        init {
            context = view.context
            media_image = view.findViewById(R.id.media_image)


            val displayMetrics = DisplayMetrics()
            (context as Activity).windowManager.defaultDisplay.getMetrics(displayMetrics)

            val width = displayMetrics.widthPixels
            val factor = 4
            media_image.layoutParams.height = width / factor

        }

        override fun bindView(item: MediaModel, payloads: List<Any>) {
            if (item.isVideoFile) {
                Glide.with(context)
                        .load(item.file)
                        .apply(RequestOptions().fitCenter().centerCrop())
                        .thumbnail(0.1f)
                        .into(media_image)
            } else {
                Glide.with(context)
                        .load(item.file)
                        .apply(RequestOptions().fitCenter().centerCrop())
                        .into(media_image)
            }
        }

        override fun unbindView(item: MediaModel) {}
    }

    override fun describeContents(): Int  = 0

    override fun writeToParcel(dest: Parcel, flags: Int) : Unit
            = dest.writeSerializable(this.file)

    protected constructor(`in`: Parcel) {
        this.file = `in`.readSerializable() as File
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<MediaModel>
                = object : Parcelable.Creator<MediaModel> {
            override fun createFromParcel(source: Parcel): MediaModel {
                return MediaModel(source)
            }

            override fun newArray(size: Int): Array<MediaModel?> {
                return arrayOfNulls(size)
            }
        }
    }
}
